### PR TITLE
Implement asm io ops for local procedure memory

### PR DESCRIPTION
**Pull request by grjte**
This PR addresses the 3rd stage of #84 which adds assembly i/o ops for local procedure variables and completes #102 by adding single-element ```push``` and ```pop``` for locals.

- updates procedure parsers to allow declaration of local variables
- adds assembly io ops for locals: ```push.local.i```,```pushw.local.i```,```pop.local.i```,```popw.local.i```,```loadw.local.i```,```storew.local.i```
- adds unit tests for local io ops and procedure parsing with locals specified and accessed
- refactors io ops module structure
- adds docs for local memory io parsers and updates RAM parser docs to address previous question

### CODE REVIEW

File: assembly/src/parsers/blocks.rs [hunk 1]
```diff
@@ -54,6 +91,7 @@ impl BlockParser {
         &self,
         tokens: &mut TokenStream,
         proc_map: &BTreeMap<String, CodeBlock>,
+        proc_locals: u32,
```
Comment (bobbinth) [inline assembly/src/parsers/blocks.rs:94]: nit: would prefer `num_proc_locals`.

File: assembly/src/parsers/blocks.rs [hunk 2]
```diff
@@ -35,6 +37,41 @@ pub fn parse_code_blocks(
...
+    let alloc_ops = vec![Operation::Push(locals_felt), Operation::FmpUpdate];
+    blocks.push(CodeBlock::new_span(alloc_ops));
+
+    // add the procedure body code block
+    blocks.push(body);
+
+    // deallocate procedure locals after the procedure body
+    let dealloc_ops = vec![
+        Operation::Push(locals_felt),
+        Operation::Neg,
```
Comment (bobbinth) [inline assembly/src/parsers/blocks.rs:66]: We could probably do this as a single operation: `Operation::Push(-locals_felt)`

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! This is great! I think the parsing of local ops is good to go (I left a few comments inline, but they are fairly minor).

I do have a question about handling updates to `fmp` though. Specifically, right now, we put the `fmp` update code inside the procedure code, and this requires creating new `Span` blocks at the begging and the end of the procedure. If the procedure inlining is done correctly, these `Span` blocks should get eliminated, but I'm not sure if it works properly in all cases now.

Another approach could injecting `fmp` update operations around the procedure call site (e.g., around `exec.<proc_name>`) - and I wonder if this would be a better approach. Specifically, this would avoid the need for creating `Span` blocks (since `exec` is already in a `Span` block). But it would also meant that:

1. We'd need to store info about procedure locals in `proc_map` (i.e., it wouldn't be just `BTreeMap<String, CodeBlock>` but maybe something like `BTreeMap<String, ProcDescriptor>` where `ProcDescriptor` contains both the number of locals and the root code block) - though, I think this might be a good idea regardless of the approach chosen.
2. We'd need to modify the `Span` block parsing logic (in `BlockParser`).  I haven't thought through what this would entail - and if the changes needed would reduce code clarity, it might not be worth it.

File: assembly/src/parsers/io_ops/local_ops.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    parse_read_local(span_ops, op, proc_locals, true)?;
+
+    for _ in 0..3 {
+        span_ops.push(Operation::Drop);
+    }
+
+    Ok(())
+}
+
+/// Pops the top element off the stack and saves it at the specified local procedure memory index.
```
[the remark this answers is not in the document]
Comment (grjte) [reply assembly/src/parsers/io_ops/local_ops.rs]: Updated!

File: assembly/src/parsers/io_ops/local_ops.rs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    // parse the provided local memory index
+    let index = parse_int_param(op, 2, 0, proc_locals - 1)?;
+
+    // put the absolute memory address on the stack
+    push_value(span_ops, Felt::new(index as u64));
+    if index != 0 {
+        // negate the value to use it as an offset from the fmp
+        // since the fmp value was updated when locals were allocated
+        span_ops.push(Operation::Neg);
+    }
```
[the remark this answers is not in the document]
Comment (grjte) [reply assembly/src/parsers/io_ops/local_ops.rs]: Yep, fixed!

File: assembly/src/parsers/io_ops/local_ops.rs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+///
+/// This function expects a memory write assembly operation that has already been validated, except
+/// for its parameter. If called without validation, it could yield incorrect results or return an
+/// `AssemblyError`. It will also return an `AssemblyError` if the op does not have an index
+/// parameter specified.
+pub fn parse_write_local(
+    span_ops: &mut Vec<Operation>,
+    op: &Token,
+    proc_locals: u32,
+    preserve_stack: bool,
```
[the remark this answers is not in the document]
Comment (grjte) [reply assembly/src/parsers/io_ops/local_ops.rs]: Happy to change this - it's clearer. Updated!

File: assembly/src/parsers/io_ops/local_ops.rs [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+///
+/// This function expects a memory read assembly operation that has already been validated, except
+/// for its parameter. If called without validation, it could yield incorrect results or return an
+/// `AssemblyError`. It will also return an `AssemblyError` if the op does not have an index
+/// parameter specified.
+pub fn parse_read_local(
+    span_ops: &mut Vec<Operation>,
+    op: &Token,
+    proc_locals: u32,
+    preserve_stack: bool,
```
[the remark this answers is not in the document]
Comment (grjte) [reply assembly/src/parsers/io_ops/local_ops.rs]: Yep, this is clearer. Updated!

File: assembly/src/parsers/blocks.rs (hunk 1 above)
Comment (grjte) [reply assembly/src/parsers/blocks.rs]: Updated throughout

File: assembly/src/parsers/blocks.rs (hunk 2 above)
Comment (grjte) [reply assembly/src/parsers/blocks.rs]: Yes, fixed!

**Comment by grjte** [comment]
> Another approach could injecting fmp update operations around the procedure call site (e.g., around exec.<proc_name>) - and I wonder if this would be a better approach.

Possibly. I agree it's worth exploring.

> We'd need to modify the Span block parsing logic (in BlockParser). I haven't thought through what this would entail - and if the changes needed would reduce code clarity, it might not be worth it.

This is basically why I didn't take this approach initially. I wanted to minimize the changes and maintain clarity. However, I think that modifying the Span block parsing logic may be required to address #109, in which case it would make sense to explore the option of moving the fmp update operations at the same time.

**Review by bobbinth — APPROVED** [review]
Everything looks good! Thank you so much!

**Pull request merged**